### PR TITLE
Fix issues related to rq job dependencies

### PR DESCRIPTION
**Pull request by Marishka17**
### Motivation and context

This PR resolves 2 issues related to rq jobs:
1. In some scenarios it is possible to reach a situation when `X2` rq job depends on `X1` -> running `X1` job is moved to FailedJobRegistry and not deleted -> user creates one more `X1` job that is enqueued after `X2` execution.  It was possible due to the second issue when a user tried to export annotations for task 1, then for task 2, and after restarting the worker container user tried again to export annotations for task 1. (cyclic dependence)
2. Looks like in rq implementation rq jobs that depend on `X` job will never be enqueued when `X` is moved to FailedJobRegistry due to AbandonedJobError. I've submitted the [issue](https://github.com/rq/rq/issues/2006) to the rq repository.
### How has this been tested?

Manually
### Checklist

- [x] I submit my changes into the `develop` branch
- [x] I have created a changelog fragment 
~~- [ ] I have updated the documentation accordingly~~
- [ ] I have added tests to cover my changes
- [ ] I have linked related issues (see [GitHub docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
~~- [ ] I have increased versions of npm packages if it is necessary
  ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
  [cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning),
  [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and
  [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))~~

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.

### CODE REVIEW

File: cvat/settings/base.py [hunk 1]
```diff
@@ -723,3 +723,66 @@ class CVAT_QUEUES(Enum):
...
+                    job.cleanup(ttl=-1, pipeline=pipeline)
+                    failed_job_registry.add(job, job.failure_ttl)
+                    queue.enqueue_dependents(job)
+
+            pipeline.zremrangebyscore(self.key, 0, score)
+            pipeline.execute()
+
+    return job_ids
+
+rq.registry.StartedJobRegistry.cleanup = custom_started_job_registry_cleanup
```
Comment (azhavoro) [inline cvat/settings/base.py:788]: Please add a comment why we need override this method, and also add a link to the relevant issue

File: cvat/apps/engine/utils.py [hunk 2]
```diff
@@ -171,9 +176,24 @@ def define_dependent_job(queue: DjangoRQ, user_id: int, should_be_dependent: boo
         for job in queue.job_class.fetch_many(
             queue.deferred_job_registry.get_job_ids(), queue.connection
         )
-        if job and job.meta.get("user", {}).get("id") == user_id
+        if job and job.meta.get("user", {}).get("id") == user_id and job.is_deferred
```
Comment (Marishka17) [inline cvat/apps/engine/utils.py:179]: We should check each job from `deferred_job_registry` that it is still deferred because `deferred_job_registry.get_job_ids()` calls `deferred_job_registry.cleanup()` which is dummy and has an "empty" implementation

File: cvat/rq_patching.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+import rq.registry
+from rq.exceptions import AbandonedJobError, NoSuchJobError
+from rq.job import JobStatus
+from rq.utils import current_timestamp
+
+
+# NOTE: we should patch implementation of original method because
+# there is no enqueuing dependent jobs in original function
+# https://github.com/rq/rq/issues/2006
```
Comment (SpecLad) [inline cvat/rq_patching.py:17]: Since you already have a fix, maybe you could submit it upstream? That way we could potentially remove this hack later (or maybe not apply it in the first place, depending on how quickly the maintainers respond).

File: cvat/rq_patching.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+                    failed_job_registry.add(job, job.failure_ttl)
+                    queue.enqueue_dependents(job)
+
+            pipeline.zremrangebyscore(self.key, 0, score)
+            pipeline.execute()
+
+    return job_ids
+
+def update_started_job_registry_cleanup() -> None:
+    rq.registry.StartedJobRegistry.cleanup = custom_started_job_registry_cleanup
```
Comment (SpecLad) [inline cvat/rq_patching.py:72]: I think you should assert on the version of rq here. That way, we'll at least remember to revise the patch if we update to a newer rq version.

File: cvat/apps/engine/utils.py (hunk 2 above)
Comment (Marishka17) [reply cvat/apps/engine/utils.py:180]: I have no exact explanation for that. Need to debug this more, but looks like it can happen if there are active or deferred jobs when restarting the worker container. I have seen several times that one job with `finished` status exists in 2 registries `DeferredJobRegistry` and `FinishedJobRegistry`. FinishedJobRegistry and  StartedJobRegistry have cleanup implementation and expired jobs will be deleted eventually. However, deferred jobs can be deleted from deferred job registry only directly like [here](https://github.com/rq/rq/blob/d11094af95188c197a5e11b1c4fa05b000485370/rq/queue.py#L1224C43-L1224C43). I'll send you screenshots with evidence.

Comment (SpecLad) [reply cvat/apps/engine/utils.py:180]: Okay. Can you summarize this issue in a code comment?

**Review by SpecLad — APPROVED** [review]
